src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void StartRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
+            CurlResponseMessage responseMessage = null;
             SafeCurlHandle requestHandle = null;
+            GCHandle requestStateHandle = new GCHandle();
+            IntPtr headerPtr = IntPtr.Zero;
+            bool referenceTaken = false;
```

Comment:
> There is a DangerousAddRef just before call to Task.Run

Ok, I see it.

> Is this AddRef needed or can the CurlSendIoctlCallback assume that the handle cannot get released since StartRequest is blocked on curl_easy_perform?

You need to make sure that something on the managed side is keeping the object alive.  If you stick with the current "easy" interface (see my previous comment about easy vs multi) and thus stick with a synchronous call, the stack of the blocked thread should be good enough to keep the object rooted, but you'd want to add a GC.KeepAlive call at the end to ensure that it's actually referenced and to prevent any optimizations that would prevent it from remaining rooted.